contrib/native/client/src/clientlib/channel.hpp

```diff
@@ -0,0 +1,236 @@
+            std::string& getHost(){return m_host;}
+            std::string& getPort(){return m_port;}
+            DrillClientError* getError(){ return m_pError;};
+
+        private:
+            void parseConnectString();
+            bool isDirectConnection();
+            bool isZookeeperConnection();
+            connectionStatus_t getDrillbitEndpointFromZk();
+            connectionStatus_t handleError(connectionStatus_t status, std::string msg);
+
+            std::string m_connectString;
+            std::string m_pathToDrill;
+            std::string m_protocol; 
+            std::string m_hostPortStr;
+            std::string m_host;
+            std::string m_port;
+
+            DrillClientError* m_pError;
+
+    };
+
+    class ChannelContext{
+        public:
+            ChannelContext(DrillUserProperties* props):m_properties(props){};
+            virtual ~ChannelContext(){};
+            const DrillUserProperties* getUserProperties() const { return m_properties;}
+        protected:
+            DrillUserProperties* m_properties;
+    };
+
+    class SSLChannelContext: public ChannelContext{
+        public:
+            static boost::asio::ssl::context::method getTlsVersion(std::string version){
+                if(version.empty()){
+                    return boost::asio::ssl::context::tlsv12;
+                } else if (version == "tlsv12") {
+                    return boost::asio::ssl::context::tlsv12;
+                } else if (version == "tlsv11") {
+                    return boost::asio::ssl::context::tlsv11;
+                } else if (version == "sslv23") {
+                    return boost::asio::ssl::context::sslv23;
+                } else if (version == "tlsv1") {
+                    return boost::asio::ssl::context::tlsv1;
+                } else if (version == "sslv3") {
+                    return boost::asio::ssl::context::sslv3;
+                } else {
+                    return boost::asio::ssl::context::tlsv12;
+                }
+            }
+
+        SSLChannelContext(DrillUserProperties *props, boost::asio::ssl::context::method tlsVersion, boost::asio::ssl::verify_mode verifyMode) :
+                ChannelContext(props),
+                m_SSLContext(tlsVersion) {
+                m_SSLContext.set_default_verify_paths();
+                m_SSLContext.set_options(
+                        boost::asio::ssl::context::default_workarounds
+                        | boost::asio::ssl::context::no_sslv2
+                        | boost::asio::ssl::context::single_dh_use
+                        );
+                m_SSLContext.set_verify_mode(verifyMode);
+            };
+            ~SSLChannelContext(){};
+            boost::asio::ssl::context& getSslContext(){ return m_SSLContext;}
+        private:
+            boost::asio::ssl::context m_SSLContext;
+    };
+
+    typedef ChannelContext ChannelContext_t; 
+    typedef SSLChannelContext SSLChannelContext_t; 
+
+    class ChannelContextFactory{
```

Comment:
I wonder about having one factory for context and one for channel, vs one factory for a given channel type (to create both channel and socket)

Reply:
Done. That got rid of a bunch of code.